File: ruby/rna-transcription/rna_transcription.rb
# class Complement
#   VERSION = 3
#   CONVERSION = {
#       'G'=>'C',
#       'C'=>'G',
#       'T'=>'A',
#       'A'=>'U'
#   }

#   def self.of_dna(dna_strand)
#     rna_strand = dna_strand.chars.map do |letter|
#       raise ArgumentError unless CONVERSION.keys.include?(letter)
#       CONVERSION[letter]
#     end
#     rna_strand.join
#   end
# end


class Complement
  VERSION = 3
  KEY = {
    'G' => 'C',
    'C' => 'G',
    'T' => 'A',
    'A' => 'U'
  }

  def self.of_dna(dna_strand)
    raise ArgumentError if dna_strand.match(/[^GCTA]/)
    dna_strand.split('').map { |dna_nuc| dna_nuc = KEY[dna_nuc] }.join
  end
end

File: javascript/rna-transcription/rna-transcription.js
var DnaTranscriber = function() {
  this.key = {
    G: 'C',
    C: 'G',
    T: 'A',
    A: 'U'
  }
};

DnaTranscriber.prototype.toRna = function(dna_strand) {
  var rna = '',
      self = this;

  dna_strand.split('').forEach(function(letter) {
   for (dna_nuc in self.key) {
     if (dna_nuc === letter) { rna += self.key[dna_nuc] };
   }
  });

  return rna;
};

module.exports = DnaTranscriber;